benchmarks/ddp/benchmark.py

```diff
@@ -0,0 +1,276 @@
+    print_measurements(prefix, benchmark.batch_size, measurements)
+    return measurements
+
+
+def sweep(benchmark):
+    # Synthesize the set of benchmarks to run.
+    # This list contain tuples for ("string prefix", [rank...]).
+    benchmarks = []
+
+    def append_benchmark(prefix, ranks):
+        prefix = "%4d GPUs -- %s" % (len(ranks), prefix)
+        benchmarks.append((prefix, ranks))
+
+    # Every process runs once by themselves to warm up (CUDA init, etc).
+    append_benchmark("  warmup", [dist.get_rank()])
+
+    # Single machine baselines
+    append_benchmark("   1M/1G", range(1))
+    append_benchmark("   1M/2G", range(2))
+    append_benchmark("   1M/4G", range(4))
+
+    # Multi-machine benchmarks
+    for i in range(1, (dist.get_world_size() // 8) + 1):
+        append_benchmark("   %dM/8G" % i, range(i * 8))
+
+    # Run benchmarks in order of increasing number of GPUs
+    results = []
+    for prefix, ranks in sorted(benchmarks, key=lambda tup: len(tup[1])):
+        # Turn range into materialized list.
+        ranks = list(ranks)
+        measurements = run_benchmark(prefix, ranks, benchmark)
+        results.append({"ranks": ranks, "measurements": measurements})
+    return results
+
+
+class Benchmark(object):
+    def __init__(self, device):
+        self.device = device
+        self.batch_size = 32
+
+    def __str__(self):
+        raise NotImplementedError
+
+    def create_model(self):
+        raise NotImplementedError
+
+    def generate_inputs(self):
+        raise NotImplementedError
+
+    def generate_target(self):
+        raise NotImplementedError
+
+
+class TorchvisionBenchmark(Benchmark):
+    def __init__(self, device, model):
+        super(TorchvisionBenchmark, self).__init__(device)
+        self.model = model
+
+    def __str__(self):
+        return "{} with batch size {}".format(self.model, self.batch_size)
+
+    def create_model(self):
+        return torchvision.models.__dict__[self.model]().to(self.device)
+
+    def generate_inputs(self):
+        return [torch.rand([self.batch_size, 3, 224, 224], device=self.device)]
+
+    def generate_target(self):
+        return torch.tensor([1] * self.batch_size, dtype=torch.long, device=self.device)
+
+
+def main():
+    parser = argparse.ArgumentParser(description='PyTorch distributed benchmark suite')
+    parser.add_argument("--rank", type=int, default=os.environ["RANK"])
+    parser.add_argument("--world-size", type=int, required=True)
+    parser.add_argument("--master-addr", type=str, required=True)
+    parser.add_argument("--master-port", type=str, required=True)
+    parser.add_argument("--model", type=str)
+    parser.add_argument("--json", type=str, metavar="PATH", help="Write file with benchmark results")
+    args = parser.parse_args()
+
+    num_gpus_per_node = torch.cuda.device_count()
+    assert num_gpus_per_node == 8, "Expected 8 GPUs per machine"
+
+    dist.init_process_group(
+        backend="gloo",
```

Comment:
Maybe we can add a comment here mentioning this process_group is just used for initialization and we don't actually use this for the actual benchmark.

Reply:
Done.